File: src/cli/features.rs
use anyhow::Result;
use clap::Subcommand;
use tabled::Tabled;

use crate::api::client::ApiClient;
use crate::api::models::{Feature, ListResponse};
use crate::cli::auth::require_auth;
use crate::utils::{output, spinner};

#[derive(Subcommand)]
pub enum FeatureCommands {
    /// List all features
    List {
        #[arg(long)]
        json: bool,
    },
    /// Get a feature by ID
    Get {
        id: String,
        #[arg(long)]
        json: bool,
    },
    /// Create a new feature from a JSON file
    Create {
        #[arg(long)]
        json: String,
    },
    /// Delete a feature by ID
    Delete { id: String },
}

#[derive(Tabled, serde::Serialize)]
struct FeatureRow {
    #[tabled(rename = "ID")]
    id: String,
    #[tabled(rename = "Name")]
    name: String,
    #[tabled(rename = "Lookup Key")]
    lookup_key: String,
    #[tabled(rename = "Type")]
    feature_type: String,
    #[tabled(rename = "Status")]
    status: String,
}

impl From<Feature> for FeatureRow {
    fn from(f: Feature) -> Self {
        Self {
            id: f.id,
            name: f.name.unwrap_or_default(),
            lookup_key: f.lookup_key.unwrap_or_default(),
            feature_type: f.feature_type.unwrap_or_default(),
            status: f.status.map(|s| output::status_badge(&s)).unwrap_or_default(),
        }
    }
}

pub async fn handle(cmd: FeatureCommands) -> Result<()> {
    let creds = require_auth()?;
    let client = ApiClient::new(creds)?;

    match cmd {
        FeatureCommands::List { json } => {
            let sp = spinner::create_spinner("Fetching features...");
            let resp: ListResponse<Feature> = client.get("/v1/features").await?;
            sp.finish_and_clear();
            let rows: Vec<FeatureRow> = resp.items.into_iter().map(Into::into).collect();
            println!("{}", output::print_table(&rows, json));
        }
        FeatureCommands::Get { id, json } => {
            let sp = spinner::create_spinner("Fetching feature...");
            let feature: Feature = client.get(&format!("/v1/features/{}", id)).await?;
            sp.finish_and_clear();
            println!("{}", output::print_detail(&feature, json));
        }
        FeatureCommands::Create { json: file } => {
            let data = std::fs::read_to_string(&file)?;
            let body: serde_json::Value = serde_json::from_str(&data)?;
            let sp = spinner::create_spinner("Creating feature...");
            let feature: Feature = client.post("/v1/features", &body).await?;
            sp.finish_and_clear();
            output::success(&format!("Feature created: {}", feature.id));
            println!("{}", output::print_detail(&feature, false));
        }
        FeatureCommands::Delete { id } => {
            let sp = spinner::create_spinner("Deleting feature...");
            client.delete_empty(&format!("/v1/features/{}", id)).await?;
            sp.finish_and_clear();
            output::success(&format!("Feature {} deleted.", id));
        }
    }
    Ok(())
}

File: npm/src/cli/features.ts
import * as fs from 'node:fs';
import { ApiClient } from '../api/client.js';
import type { Feature, ListResponse } from '../api/models.js';
import * as output from '../utils/output.js';
import { createSpinner } from '../utils/spinner.js';
import { requireAuth } from './auth.js';
import { requiredInput, select, toSnakeCase } from './prompts.js';
import type { GlobalOpts } from './types.js';

const FEATURE_TYPE_CHOICES = [
  { name: 'BOOLEAN — on/off feature flag', value: 'BOOLEAN' },
  { name: 'METERED — usage-based feature', value: 'METERED' },
  { name: 'STATIC — static value feature', value: 'STATIC' },
];

export async function featuresList(globalOpts: GlobalOpts, json: boolean): Promise<void> {
  const creds = requireAuth(globalOpts);
  const client = new ApiClient(creds);
  const sp = createSpinner('Fetching features...');
  const resp = (await client.get('/v1/features')) as ListResponse<Feature>;
  sp.stop();
  const rows = (resp.items ?? []).map((f) => ({
    ID: f.id ?? '',
    Name: f.name ?? '',
    'Lookup Key': f.lookup_key ?? '',
    Type: f.type ?? '',
    Status: f.status ? output.statusBadge(f.status) : '',
  }));
  process.stdout.write(output.printTable(rows, json));
}

export async function featuresGet(globalOpts: GlobalOpts, id: string, json: boolean): Promise<void> {
  const creds = requireAuth(globalOpts);
  const client = new ApiClient(creds);
  const sp = createSpinner('Fetching feature...');
  const feature = (await client.get(`/v1/features/${id}`)) as Feature;
  sp.stop();
  process.stdout.write(output.printDetail(feature, json));
}

export async function featuresCreate(globalOpts: GlobalOpts, file?: string): Promise<void> {
  const creds = requireAuth(globalOpts);
  const client = new ApiClient(creds);

  let body: unknown;
  if (file) {
    body = JSON.parse(fs.readFileSync(file, 'utf-8'));
  } else {
    output.info('Create a new feature');
    const name = await requiredInput('Feature name:');
    const lookup_key = await requiredInput(`Lookup key (suggested: ${toSnakeCase(name)}):`);
    const type = await select({
      message: 'Feature type:',
      choices: FEATURE_TYPE_CHOICES,
    });
    body = { name, lookup_key: lookup_key || toSnakeCase(name), type };
  }

  const sp = createSpinner('Creating feature...');
  const feature = (await client.post('/v1/features', body)) as Feature;
  sp.stop();
  output.success(`Feature created: ${feature.id ?? ''}`);
  process.stdout.write(output.printDetail(feature, false));
}

export async function featuresDelete(globalOpts: GlobalOpts, id: string): Promise<void> {
  const creds = requireAuth(globalOpts);
  const client = new ApiClient(creds);
  const sp = createSpinner('Deleting feature...');
  await client.deleteEmpty(`/v1/features/${id}`);
  sp.stop();
  output.success(`Feature ${id} deleted.`);
}
